src/waveresponse/_core.py

```diff
@@ -884,6 +884,89 @@ def from_grid(cls, grid):
         )
 
+    @classmethod
+    def from_spectrum1d(
+        cls,
+        freq,
+        dirs,
+        spectrum1d,
+        spread_fun,
+        dirp,
+        freq_hz=False,
+        degrees=False,
+        clockwise=False,
+        waves_coming_from=True,
+    ):
+        """
+        Construct a 2-D 'directional' spectrum from a 1-D 'non-directional' spectrum,
+        a spreading function and a peak direction.
+
+        The directional spectrum is constructed according to:
+
+            ``S(f, theta) = S(f) * D(f, theta - theta_p)``
+
+        where ``S(f)`` is the non-directional spectrum, ``D(f, theta - theta_p)``
+        is the directional spreading function, and ``theta_p`` is the peak direction.
+        ``f`` is the frequency coordinate, and ``theta`` is the direction coordinate.
+
+        Parameters
+        ----------
+        freq : array-like
+            1-D array of grid frequency coordinates. Positive and monotonically increasing.
+        dirs : array-like
+            1-D array of grid direction coordinates. Positive and monotonically increasing.
+            Must cover the directional range [0, 360) degrees (or [0, 2 * numpy.pi) radians).
+        spectrum1d : array-like
+            1-D array of non-directional spectrum density values. These 1-D spectrum
+            values will be scaled according to the spreading function, and distributed
+            to all frequency/direction coordinates. `spectrum1d` must have the same
+            length as `freq`.
+        spread_fun : callable
+            Spreading function. Takes a frequency coordinate (float) and a direction
+            coordinate (float) as input, and returns a corresponding scaling value
+            (float).
+        dirp : float
+            Peak direction. Direction in which the spectrum has its maximum values.
+        freq_hz : bool
+            If frequency is given in 'Hz'. If ``False``, 'rad/s' is assumed.
+        degrees : bool
+            If direction is given in 'degrees'. If ``False``, 'radians' is assumed.
+        clockwise : bool
+            If positive directions are defined to be 'clockwise'. If ``False``,
+            'counterclockwise' is assumed.
+        waves_coming_from : bool
+            If waves are 'coming from' the given directions. If ``False``, 'going towards'
+            convention is assumed.
+        """
+        freq = np.asarray_chkfinite(freq)
+        dirs = np.asarray_chkfinite(dirs)
+        spectrum1d = np.asarray_chkfinite(spectrum1d)
+        vals = (
+            np.repeat(spectrum1d, len(dirs))
+            .reshape(len(freq), len(dirs))
+            .astype("float64")
+        )
+
+        if degrees:
+            period = 360.0
+        else:
+            period = 2.0 * np.pi
+
+        for (idx_f, idx_d), val_i in np.ndenumerate(vals):
+            f_i = freq[idx_f]
+            d_i = (dirs[idx_d] - dirp) % period
+            vals[idx_f, idx_d] = spread_fun(f_i, d_i) * val_i
```

Comment:
```suggestion
        dirs_spread = (dirs - dirp) % period
        for (idx_f, idx_d), val_i in np.ndenumerate(vals):
            vals[idx_f, idx_d] = spread_fun(freq[idx_f], dirs_spread[idx_d]) * val_i
```
Small vectorization... Effect is probably minimal. More readible?